cg/services/deliver_files/file_formatter/component_files/concatenation_service.py

```diff
@@ -0,0 +1,307 @@
+            )
+        return concatenation_map
+
+    @staticmethod
+    def _determine_read_direction(fastq_path: Path) -> ReadDirection:
+        """Determine the read direction of a fastq file.
+            Assumes that the fastq file path contains 'R1' or 'R2' to determine the read direction.
+        args:
+            fastq_path: Path: Path to the fastq file.
+        """
+        if f"R{ReadDirection.FORWARD}" in fastq_path.as_posix():
+            return ReadDirection.FORWARD
+        return ReadDirection.REVERSE
+
+    def _group_fastq_files_per_sample(
+        self, sample_names: set[str], fastq_files: list[FastqFile]
+    ) -> dict[str, list[FastqFile]]:
+        """Group fastq files per sample.
+        returns a dictionary with sample names as keys and a list of fastq files as values.
+        args:
+            sample_names: set[str]: Set of sample names.
+            fastq_files: list[FastqFile]: List of fastq files.
+        """
+
+        sample_fastq_files: dict[str, list[FastqFile]] = {
+            sample_name: [] for sample_name in sample_names
+        }
+        for fastq_file in fastq_files:
+            sample_fastq_files[fastq_file.sample_name].append(fastq_file)
+        self._all_sample_fastq_file_share_same_directory(sample_fastq_files=sample_fastq_files)
+        return sample_fastq_files
+
+    def _replace_fastq_paths(
+        self,
+        concatenation_maps: dict[Path, Path],
+        formatted_files: list[FormattedFile],
+    ) -> None:
+        """
+        Replace the fastq file paths with the new concatenated fastq file paths.
+        Uses the concatenation map with the formatted file path as key and the concatenated path as value.
+        args:
+            concatenation_maps: list[ConcatenationMap]: List of ConcatenationMap objects.
+            formatted_files: list[FormattedFile]: List of formatted files.
+        """
+        for formatted_file in formatted_files:
+            if self._is_lane_fastq_file(formatted_file.formatted_path):
+                formatted_file.formatted_path = concatenation_maps[formatted_file.formatted_path]
+
+    @staticmethod
+    def _all_sample_fastq_file_share_same_directory(
+        sample_fastq_files: dict[str, list[FastqFile]]
+    ) -> None:
+        """
+        Assert that all fastq files for a sample share the same directory.
+        This is to ensure that the files are concatenated within the expected directory path.
+        args:
+            sample_fastq_files: dict[str, list[FastqFile]]: Dictionary of sample names and their fastq files.
+        """
+        for sample_name in sample_fastq_files.keys():
+            fastq_files: list[FastqFile] = sample_fastq_files[sample_name]
+            parent_dir: Path = fastq_files[0].fastq_file_path.parent
+            for fastq_file in fastq_files:
+                if fastq_file.fastq_file_path.parent != parent_dir:
+                    raise ValueError(
```

Comment:
Make sure this error is caught somewhere and we know what to expect when this fails. I could not see any error handling for this in the fohm upload

Reply:
the whole process should fail if this error happens as it signals we are trying to do something that isn't possible. Or do you want to catch this and re-raise a more specific error?